src/partisan_hyparview_peer_service_manager.erl

```diff
@@ -507,37 +526,57 @@ code_change(_OldVsn, State, _Extra) ->
 %%%===================================================================
 
 %% @private
-handle_message({disconnect, Peer}, #state{myself=Myself,
-                                          active=Active0}=State0) ->
-    lager:info("Node ~p received disconnect from ~p", [Myself, Peer]),
+handle_message({disconnect, Peer, DisconnectId},
+               #state{myself=Myself,
+                      active=Active0,
+                      ack_message_map=AckMessageMap0}=State0) ->
+    lager:info("Node ~p received disconnect from ~p with ~p",
+               [Myself, Peer, DisconnectId]),
 
-    %% Remove from active
-    Active = sets:del_element(Peer, Active0),
-    lager:info("Node ~p active view: ~p", [Myself, sets:to_list(Active)]),
+    case is_valid_disconnect(Peer, DisconnectId, AckMessageMap0) of
+        false ->
+            %% Ignore the older disconnect message.
+            {reply, ok, State0};
+        true ->
+            %% Remove from active
+            Active = sets:del_element(Peer, Active0),
+            lager:info("Node ~p active view: ~p", [Myself, sets:to_list(Active)]),
 
-    %% Add to passive view.
-    State = add_to_passive_view(Peer,
-                                State0#state{active=Active}),
+            %% Add to passive view.
+            State = add_to_passive_view(Peer,
+                                        State0#state{active=Active}),
 
-    {reply, ok, State#state{active=Active}};
+            %% Update the AckMessageMap.
+            AckMessageMap = dict:store(Peer, DisconnectId, AckMessageMap0),
 
-handle_message({neighbor, Peer, TheirTag, _Sender}, State0) ->
+            {reply, ok, State#state{ack_message_map=AckMessageMap}}
+    end;
+
+handle_message({neighbor, Peer, TheirTag, DisconnectId, _Sender},
+               #state{sent_message_map=SentMessageMap}=State0) ->
     lager:info("Neighboring with ~p tagged ~p from random walk", [Peer, TheirTag]),
 
-    %% Add node into the active view.
-    State = add_to_active_view(Peer, TheirTag, State0),
+    State = case is_addable(DisconnectId, Peer, SentMessageMap) of
+                true ->
+                    %% Add node into the active view.
+                    add_to_active_view(Peer, TheirTag, State0);
+                false ->
+                    %% @todo Need to send the disconnect message here?
```

Comment:
Before sending the neighbor message, the functions (like `perform_join()`) add the peer to their active view. So, if the neighbor message fails to add the peer, I think we need to let the peer know that neighboring was failed.